// three.cjs/index.js
var Three = require('./Three');
exports.REVISION = Three.REVISION;
exports.log = Three.log;
exports.warn = Three.warn;
exports.error = Three.error;
exports.MOUSE = Three.MOUSE;
exports.CullFaceNone = Three.CullFaceNone;
exports.CullFaceBack = Three.CullFaceBack;
exports.CullFaceFront = Three.CullFaceFront;
exports.CullFaceFrontBack = Three.CullFaceFrontBack;
exports.FrontFaceDirectionCW = Three.FrontFaceDirectionCW;
exports.FrontFaceDirectionCCW = Three.FrontFaceDirectionCCW;
exports.BasicShadowMap = Three.BasicShadowMap;
exports.PCFShadowMap = Three.PCFShadowMap;
exports.PCFSoftShadowMap = Three.PCFSoftShadowMap;
exports.FrontSide = Three.FrontSide;
exports.BackSide = Three.BackSide;
exports.DoubleSide = Three.DoubleSide;
exports.NoShading = Three.NoShading;
exports.FlatShading = Three.FlatShading;
exports.SmoothShading = Three.SmoothShading;
exports.NoColors = Three.NoColors;
exports.FaceColors = Three.FaceColors;
exports.VertexColors = Three.VertexColors;
exports.NoBlending = Three.NoBlending;
exports.NormalBlending = Three.NormalBlending;
exports.AdditiveBlending = Three.AdditiveBlending;
exports.SubtractiveBlending = Three.SubtractiveBlending;
exports.MultiplyBlending = Three.MultiplyBlending;
exports.CustomBlending = Three.CustomBlending;
exports.AddEquation = Three.AddEquation;
exports.SubtractEquation = Three.SubtractEquation;
exports.ReverseSubtractEquation = Three.ReverseSubtractEquation;
exports.MinEquation = Three.MinEquation;
exports.MaxEquation = Three.MaxEquation;
exports.ZeroFactor = Three.ZeroFactor;
exports.OneFactor = Three.OneFactor;
exports.SrcColorFactor = Three.SrcColorFactor;
exports.OneMinusSrcColorFactor = Three.OneMinusSrcColorFactor;
exports.SrcAlphaFactor = Three.SrcAlphaFactor;
exports.OneMinusSrcAlphaFactor = Three.OneMinusSrcAlphaFactor;
exports.DstAlphaFactor = Three.DstAlphaFactor;
exports.OneMinusDstAlphaFactor = Three.OneMinusDstAlphaFactor;
exports.DstColorFactor = Three.DstColorFactor;
exports.OneMinusDstColorFactor = Three.OneMinusDstColorFactor;
exports.SrcAlphaSaturateFactor = Three.SrcAlphaSaturateFactor;
exports.MultiplyOperation = Three.MultiplyOperation;
exports.MixOperation = Three.MixOperation;
exports.AddOperation = Three.AddOperation;
exports.UVMapping = Three.UVMapping;
exports.CubeReflectionMapping = Three.CubeReflectionMapping;
exports.CubeRefractionMapping = Three.CubeRefractionMapping;
exports.EquirectangularReflectionMapping = Three.EquirectangularReflectionMapping;
exports.EquirectangularRefractionMapping = Three.EquirectangularRefractionMapping;
exports.SphericalReflectionMapping = Three.SphericalReflectionMapping;
exports.RepeatWrapping = Three.RepeatWrapping;
exports.ClampToEdgeWrapping = Three.ClampToEdgeWrapping;
exports.MirroredRepeatWrapping = Three.MirroredRepeatWrapping;
exports.NearestFilter = Three.NearestFilter;
exports.NearestMipMapNearestFilter = Three.NearestMipMapNearestFilter;
exports.NearestMipMapLinearFilter = Three.NearestMipMapLinearFilter;
exports.LinearFilter = Three.LinearFilter;
exports.LinearMipMapNearestFilter = Three.LinearMipMapNearestFilter;
exports.LinearMipMapLinearFilter = Three.LinearMipMapLinearFilter;
exports.UnsignedByteType = Three.UnsignedByteType;
exports.ByteType = Three.ByteType;
exports.ShortType = Three.ShortType;
exports.UnsignedShortType = Three.UnsignedShortType;
exports.IntType = Three.IntType;
exports.UnsignedIntType = Three.UnsignedIntType;
exports.FloatType = Three.FloatType;
exports.HalfFloatType = Three.HalfFloatType;
exports.UnsignedShort4444Type = Three.UnsignedShort4444Type;
exports.UnsignedShort5551Type = Three.UnsignedShort5551Type;
exports.UnsignedShort565Type = Three.UnsignedShort565Type;
exports.AlphaFormat = Three.AlphaFormat;
exports.RGBFormat = Three.RGBFormat;
exports.RGBAFormat = Three.RGBAFormat;
exports.LuminanceFormat = Three.LuminanceFormat;
exports.LuminanceAlphaFormat = Three.LuminanceAlphaFormat;
exports.RGBEFormat = Three.RGBEFormat;
exports.RGB_S3TC_DXT1_Format = Three.RGB_S3TC_DXT1_Format;
exports.RGBA_S3TC_DXT1_Format = Three.RGBA_S3TC_DXT1_Format;
exports.RGBA_S3TC_DXT3_Format = Three.RGBA_S3TC_DXT3_Format;
exports.RGBA_S3TC_DXT5_Format = Three.RGBA_S3TC_DXT5_Format;
exports.RGB_PVRTC_4BPPV1_Format = Three.RGB_PVRTC_4BPPV1_Format;
exports.RGB_PVRTC_2BPPV1_Format = Three.RGB_PVRTC_2BPPV1_Format;
exports.RGBA_PVRTC_4BPPV1_Format = Three.RGBA_PVRTC_4BPPV1_Format;
exports.RGBA_PVRTC_2BPPV1_Format = Three.RGBA_PVRTC_2BPPV1_Format;
exports.GeometryIdCount = Three.GeometryIdCount;
exports.Object3DIdCount = Three.Object3DIdCount;
exports.MaterialIdCount = Three.MaterialIdCount;
exports.TextureIdCount = Three.TextureIdCount;
exports.LineStrip = Three.LineStrip;
exports.LinePieces = Three.LinePieces;
exports.Projector = Three.Projector;
exports.CanvasRenderer = Three.CanvasRenderer;
exports.Camera = require('./cameras/Camera').Camera;
exports.CubeCamera = require('./cameras/CubeCamera').CubeCamera;
exports.OrthographicCamera = require('./cameras/OrthographicCamera').OrthographicCamera;
exports.PerspectiveCamera = require('./cameras/PerspectiveCamera').PerspectiveCamera;
var BufferAttribute = require('./core/BufferAttribute');
exports.BufferAttribute = BufferAttribute.BufferAttribute;
exports.Int8Attribute = BufferAttribute.Int8Attribute;
exports.Uint8Attribute = BufferAttribute.Uint8Attribute;
exports.Uint8ClampedAttribute = BufferAttribute.Uint8ClampedAttribute;
exports.Int16Attribute = BufferAttribute.Int16Attribute;
exports.Uint16Attribute = BufferAttribute.Uint16Attribute;
exports.Int32Attribute = BufferAttribute.Int32Attribute;
exports.Uint32Attribute = BufferAttribute.Uint32Attribute;
exports.Float32Attribute = BufferAttribute.Float32Attribute;
exports.Float64Attribute = BufferAttribute.Float64Attribute;
exports.Clock = require('./core/Clock').Clock;
exports.BufferGeometry = require('./core/BufferGeometry').BufferGeometry;
exports.DynamicBufferAttribute = require('./core/DynamicBufferAttribute').DynamicBufferAttribute;
exports.Face3 = require('./core/Face3').Face3;
exports.EventDispatcher = require('./core/EventDispatcher').EventDispatcher;
exports.Face4 = require('./core/Face4').Face4;
exports.Geometry = require('./core/Geometry').Geometry;
exports.Object3D = require('./core/Object3D').Object3D;
exports.Raycaster = require('./core/Raycaster').Raycaster;
exports.CurveUtils = require('./extras/CurveUtils').CurveUtils;
var FontUtils = require('./extras/FontUtils');
exports.FontUtils = FontUtils.FontUtils;
exports.typeface_js = FontUtils.typeface_js;
exports.GeometryUtils = require('./extras/GeometryUtils').GeometryUtils;
exports.ImageUtils = require('./extras/ImageUtils').ImageUtils;
exports.SceneUtils = require('./extras/SceneUtils').SceneUtils;
exports.ShapeUtils = require('./extras/ShapeUtils').ShapeUtils;
exports.AmbientLight = require('./lights/AmbientLight').AmbientLight;
var Triangulate = require('./extras/Triangulate');
exports.Triangulate = Triangulate.Triangulate;
exports.TriangulateArea = Triangulate.TriangulateArea;
exports.AreaLight = require('./lights/AreaLight').AreaLight;
exports.DirectionalLight = require('./lights/DirectionalLight').DirectionalLight;
exports.Light = require('./lights/Light').Light;
exports.HemisphereLight = require('./lights/HemisphereLight').HemisphereLight;
exports.PointLight = require('./lights/PointLight').PointLight;
exports.SpotLight = require('./lights/SpotLight').SpotLight;
var BinaryTextureLoader = require('./loaders/BinaryTextureLoader');
exports.DataTextureLoader = BinaryTextureLoader.DataTextureLoader;
exports.BinaryTextureLoader = BinaryTextureLoader.BinaryTextureLoader;
exports.BufferGeometryLoader = require('./loaders/BufferGeometryLoader').BufferGeometryLoader;
exports.Cache = require('./loaders/Cache').Cache;
exports.CompressedTextureLoader = require('./loaders/CompressedTextureLoader').CompressedTextureLoader;
exports.GeometryLoader = require('./loaders/GeometryLoader').GeometryLoader;
exports.ImageLoader = require('./loaders/ImageLoader').ImageLoader;
exports.JSONLoader = require('./loaders/JSONLoader').JSONLoader;
exports.Loader = require('./loaders/Loader').Loader;
exports.MaterialLoader = require('./loaders/MaterialLoader').MaterialLoader;
var LoadingManager = require('./loaders/LoadingManager');
exports.LoadingManager = LoadingManager.LoadingManager;
exports.DefaultLoadingManager = LoadingManager.DefaultLoadingManager;
exports.TextureLoader = require('./loaders/TextureLoader').TextureLoader;
exports.XHRLoader = require('./loaders/XHRLoader').XHRLoader;
exports.LineBasicMaterial = require('./materials/LineBasicMaterial').LineBasicMaterial;
exports.ObjectLoader = require('./loaders/ObjectLoader').ObjectLoader;
exports.LineDashedMaterial = require('./materials/LineDashedMaterial').LineDashedMaterial;
exports.Material = require('./materials/Material').Material;
exports.MeshBasicMaterial = require('./materials/MeshBasicMaterial').MeshBasicMaterial;
exports.MeshDepthMaterial = require('./materials/MeshDepthMaterial').MeshDepthMaterial;
exports.MeshLambertMaterial = require('./materials/MeshLambertMaterial').MeshLambertMaterial;
exports.MeshFaceMaterial = require('./materials/MeshFaceMaterial').MeshFaceMaterial;
exports.MeshNormalMaterial = require('./materials/MeshNormalMaterial').MeshNormalMaterial;
exports.MeshPhongMaterial = require('./materials/MeshPhongMaterial').MeshPhongMaterial;
var PointCloudMaterial = require('./materials/PointCloudMaterial');
exports.PointCloudMaterial = PointCloudMaterial.PointCloudMaterial;
exports.ParticleBasicMaterial = PointCloudMaterial.ParticleBasicMaterial;
exports.ParticleSystemMaterial = PointCloudMaterial.ParticleSystemMaterial;
exports.RawShaderMaterial = require('./materials/RawShaderMaterial').RawShaderMaterial;
exports.ShaderMaterial = require('./materials/ShaderMaterial').ShaderMaterial;
exports.SpriteMaterial = require('./materials/SpriteMaterial').SpriteMaterial;
exports.Box2 = require('./math/Box2').Box2;
exports.Box3 = require('./math/Box3').Box3;
var Color = require('./math/Color');
exports.Color = Color.Color;
exports.ColorKeywords = Color.ColorKeywords;
exports.Euler = require('./math/Euler').Euler;
exports.Frustum = require('./math/Frustum').Frustum;
exports.Line3 = require('./math/Line3').Line3;
exports.Math = require('./math/Math').Math;
exports.Matrix3 = require('./math/Matrix3').Matrix3;
exports.Plane = require('./math/Plane').Plane;
exports.Matrix4 = require('./math/Matrix4').Matrix4;
exports.Quaternion = require('./math/Quaternion').Quaternion;
exports.Ray = require('./math/Ray').Ray;
exports.Sphere = require('./math/Sphere').Sphere;
exports.Spline = require('./math/Spline').Spline;
exports.Vector2 = require('./math/Vector2').Vector2;
exports.Triangle = require('./math/Triangle').Triangle;
exports.Vector4 = require('./math/Vector4').Vector4;
exports.Bone = require('./objects/Bone').Bone;
exports.Vector3 = require('./math/Vector3').Vector3;
exports.Group = require('./objects/Group').Group;
exports.LensFlare = require('./objects/LensFlare').LensFlare;
exports.Line = require('./objects/Line').Line;
exports.LOD = require('./objects/LOD').LOD;
exports.MorphAnimMesh = require('./objects/MorphAnimMesh').MorphAnimMesh;
exports.Mesh = require('./objects/Mesh').Mesh;
var PointCloud = require('./objects/PointCloud');
exports.PointCloud = PointCloud.PointCloud;
exports.ParticleSystem = PointCloud.ParticleSystem;
exports.Skeleton = require('./objects/Skeleton').Skeleton;
exports.SkinnedMesh = require('./objects/SkinnedMesh').SkinnedMesh;
var Sprite = require('./objects/Sprite');
exports.Sprite = Sprite.Sprite;
exports.Particle = Sprite.Particle;
exports.WebGLRenderer = require('./renderers/WebGLRenderer').WebGLRenderer;
exports.WebGLRenderTarget = require('./renderers/WebGLRenderTarget').WebGLRenderTarget;
exports.WebGLRenderTargetCube = require('./renderers/WebGLRenderTargetCube').WebGLRenderTargetCube;
exports.Fog = require('./scenes/Fog').Fog;
exports.Scene = require('./scenes/Scene').Scene;
exports.FogExp2 = require('./scenes/FogExp2').FogExp2;
exports.CompressedTexture = require('./textures/CompressedTexture').CompressedTexture;
exports.CubeTexture = require('./textures/CubeTexture').CubeTexture;
exports.DataTexture = require('./textures/DataTexture').DataTexture;
exports.Texture = require('./textures/Texture').Texture;
exports.VideoTexture = require('./textures/VideoTexture').VideoTexture;
exports.AnimationHandler = require('./extras/animation/AnimationHandler').AnimationHandler;
exports.Animation = require('./extras/animation/Animation').Animation;
exports.MorphAnimation = require('./extras/animation/MorphAnimation').MorphAnimation;
exports.KeyFrameAnimation = require('./extras/animation/KeyFrameAnimation').KeyFrameAnimation;
exports.Audio = require('./extras/audio/Audio').Audio;
exports.AudioListener = require('./extras/audio/AudioListener').AudioListener;
exports.Curve = require('./extras/core/Curve').Curve;
exports.Gyroscope = require('./extras/core/Gyroscope').Gyroscope;
exports.CurvePath = require('./extras/core/CurvePath').CurvePath;
var Path = require('./extras/core/Path');
exports.Path = Path.Path;
exports.PathActions = Path.PathActions;
exports.ArcCurve = require('./extras/curves/ArcCurve').ArcCurve;
exports.Shape = require('./extras/core/Shape').Shape;
exports.ClosedSplineCurve3 = require('./extras/curves/ClosedSplineCurve3').ClosedSplineCurve3;
exports.CubicBezierCurve = require('./extras/curves/CubicBezierCurve').CubicBezierCurve;
exports.EllipseCurve = require('./extras/curves/EllipseCurve').EllipseCurve;
exports.CubicBezierCurve3 = require('./extras/curves/CubicBezierCurve3').CubicBezierCurve3;
exports.LineCurve = require('./extras/curves/LineCurve').LineCurve;
exports.LineCurve3 = require('./extras/curves/LineCurve3').LineCurve3;
exports.QuadraticBezierCurve = require('./extras/curves/QuadraticBezierCurve').QuadraticBezierCurve;
exports.QuadraticBezierCurve3 = require('./extras/curves/QuadraticBezierCurve3').QuadraticBezierCurve3;
exports.SplineCurve = require('./extras/curves/SplineCurve').SplineCurve;
exports.SplineCurve3 = require('./extras/curves/SplineCurve3').SplineCurve3;
exports.BoxGeometry = require('./extras/geometries/BoxGeometry').BoxGeometry;
exports.CircleGeometry = require('./extras/geometries/CircleGeometry').CircleGeometry;
exports.CubeGeometry = require('./extras/geometries/CubeGeometry').CubeGeometry;
exports.CylinderGeometry = require('./extras/geometries/CylinderGeometry').CylinderGeometry;
exports.DodecahedronGeometry = require('./extras/geometries/DodecahedronGeometry').DodecahedronGeometry;
exports.ExtrudeGeometry = require('./extras/geometries/ExtrudeGeometry').ExtrudeGeometry;
exports.OctahedronGeometry = require('./extras/geometries/OctahedronGeometry').OctahedronGeometry;
exports.IcosahedronGeometry = require('./extras/geometries/IcosahedronGeometry').IcosahedronGeometry;
exports.LatheGeometry = require('./extras/geometries/LatheGeometry').LatheGeometry;
exports.ParametricGeometry = require('./extras/geometries/ParametricGeometry').ParametricGeometry;
exports.PlaneGeometry = require('./extras/geometries/PlaneGeometry').PlaneGeometry;
exports.PlaneBufferGeometry = require('./extras/geometries/PlaneBufferGeometry').PlaneBufferGeometry;
exports.PolyhedronGeometry = require('./extras/geometries/PolyhedronGeometry').PolyhedronGeometry;
exports.RingGeometry = require('./extras/geometries/RingGeometry').RingGeometry;
exports.ShapeGeometry = require('./extras/geometries/ShapeGeometry').ShapeGeometry;
exports.TetrahedronGeometry = require('./extras/geometries/TetrahedronGeometry').TetrahedronGeometry;
exports.SphereGeometry = require('./extras/geometries/SphereGeometry').SphereGeometry;
exports.TextGeometry = require('./extras/geometries/TextGeometry').TextGeometry;
exports.TorusGeometry = require('./extras/geometries/TorusGeometry').TorusGeometry;
exports.TorusKnotGeometry = require('./extras/geometries/TorusKnotGeometry').TorusKnotGeometry;
exports.TubeGeometry = require('./extras/geometries/TubeGeometry').TubeGeometry;
exports.ArrowHelper = require('./extras/helpers/ArrowHelper').ArrowHelper;
exports.BoundingBoxHelper = require('./extras/helpers/BoundingBoxHelper').BoundingBoxHelper;
exports.AxisHelper = require('./extras/helpers/AxisHelper').AxisHelper;
exports.BoxHelper = require('./extras/helpers/BoxHelper').BoxHelper;
exports.CameraHelper = require('./extras/helpers/CameraHelper').CameraHelper;
exports.DirectionalLightHelper = require('./extras/helpers/DirectionalLightHelper').DirectionalLightHelper;
exports.EdgesHelper = require('./extras/helpers/EdgesHelper').EdgesHelper;
exports.FaceNormalsHelper = require('./extras/helpers/FaceNormalsHelper').FaceNormalsHelper;
exports.GridHelper = require('./extras/helpers/GridHelper').GridHelper;
exports.HemisphereLightHelper = require('./extras/helpers/HemisphereLightHelper').HemisphereLightHelper;
exports.PointLightHelper = require('./extras/helpers/PointLightHelper').PointLightHelper;
exports.SkeletonHelper = require('./extras/helpers/SkeletonHelper').SkeletonHelper;
exports.SpotLightHelper = require('./extras/helpers/SpotLightHelper').SpotLightHelper;
exports.VertexNormalsHelper = require('./extras/helpers/VertexNormalsHelper').VertexNormalsHelper;
exports.VertexTangentsHelper = require('./extras/helpers/VertexTangentsHelper').VertexTangentsHelper;
exports.WireframeHelper = require('./extras/helpers/WireframeHelper').WireframeHelper;
exports.ImmediateRenderObject = require('./extras/objects/ImmediateRenderObject').ImmediateRenderObject;
exports.MorphBlendMesh = require('./extras/objects/MorphBlendMesh').MorphBlendMesh;
exports.ShaderChunk = require('./renderers/shaders/ShaderChunk').ShaderChunk;
exports.ShaderLib = require('./renderers/shaders/ShaderLib').ShaderLib;
exports.UniformsLib = require('./renderers/shaders/UniformsLib').UniformsLib;
exports.UniformsUtils = require('./renderers/shaders/UniformsUtils').UniformsUtils;
exports.WebGLExtensions = require('./renderers/webgl/WebGLExtensions').WebGLExtensions;
exports.WebGLProgram = require('./renderers/webgl/WebGLProgram').WebGLProgram;
exports.WebGLShader = require('./renderers/webgl/WebGLShader').WebGLShader;
exports.WebGLState = require('./renderers/webgl/WebGLState').WebGLState;
exports.WebGLTextures = require('./renderers/webgl/WebGLTextures').WebGLTextures;
exports.ShadowMapPlugin = require('./renderers/webgl/plugins/ShadowMapPlugin').ShadowMapPlugin;
exports.LensFlarePlugin = require('./renderers/webgl/plugins/LensFlarePlugin').LensFlarePlugin;
exports.SpritePlugin = require('./renderers/webgl/plugins/SpritePlugin').SpritePlugin;
