// three.cjs/extras/helpers/BoundingBoxHelper.js
/**
 * @author WestLangley / http://github.com/WestLangley
 */
// a helper to show the world-axis-aligned bounding box for an object
var MeshBasicMaterialModule = require('../../materials/MeshBasicMaterial');
var BoxGeometryModule = require('../geometries/BoxGeometry');
var MeshModule = require('../../objects/Mesh');
var Box3Module = require('../../math/Box3');
var BoundingBoxHelper;
BoundingBoxHelper = function (object, hex) {
	var color = hex !== undefined ? hex : 8947848;
	this.object = object;
	this.box = new Box3Module.Box3();
	MeshModule.Mesh.call(this, new BoxGeometryModule.BoxGeometry(1, 1, 1), new MeshBasicMaterialModule.MeshBasicMaterial({
		color: color,
		wireframe: true
	}));
};
BoundingBoxHelper.prototype = Object.create(MeshModule.Mesh.prototype);
BoundingBoxHelper.prototype.constructor = BoundingBoxHelper;
BoundingBoxHelper.prototype.update = function () {
	this.box.setFromObject(this.object);
	this.box.size(this.scale);
	this.box.center(this.position);
};
BoundingBoxHelper.prototype.isBoundingBoxHelper = true;
exports.BoundingBoxHelper = BoundingBoxHelper;
