// three.cjs/renderers/shaders/UniformsLib.js
/**
 * Uniforms library for shared webgl shaders
 */
var Vector2Module = require('../../math/Vector2');
var Vector4Module = require('../../math/Vector4');
var ColorModule = require('../../math/Color');
var UniformsLib;
UniformsLib = {
	common: {
		'diffuse': {
			type: 'c',
			value: new ColorModule.Color(15658734)
		},
		'opacity': {
			type: 'f',
			value: 1
		},
		'map': {
			type: 't',
			value: null
		},
		'offsetRepeat': {
			type: 'v4',
			value: new Vector4Module.Vector4(0, 0, 1, 1)
		},
		'lightMap': {
			type: 't',
			value: null
		},
		'specularMap': {
			type: 't',
			value: null
		},
		'alphaMap': {
			type: 't',
			value: null
		},
		'envMap': {
			type: 't',
			value: null
		},
		'flipEnvMap': {
			type: 'f',
			value: -1
		},
		'reflectivity': {
			type: 'f',
			value: 1
		},
		'refractionRatio': {
			type: 'f',
			value: 0.98
		},
		'morphTargetInfluences': {
			type: 'f',
			value: 0
		}
	},
	bump: {
		'bumpMap': {
			type: 't',
			value: null
		},
		'bumpScale': {
			type: 'f',
			value: 1
		}
	},
	normalmap: {
		'normalMap': {
			type: 't',
			value: null
		},
		'normalScale': {
			type: 'v2',
			value: new Vector2Module.Vector2(1, 1)
		}
	},
	fog: {
		'fogDensity': {
			type: 'f',
			value: 0.00025
		},
		'fogNear': {
			type: 'f',
			value: 1
		},
		'fogFar': {
			type: 'f',
			value: 2000
		},
		'fogColor': {
			type: 'c',
			value: new ColorModule.Color(16777215)
		}
	},
	lights: {
		'ambientLightColor': {
			type: 'fv',
			value: []
		},
		'directionalLightDirection': {
			type: 'fv',
			value: []
		},
		'directionalLightColor': {
			type: 'fv',
			value: []
		},
		'hemisphereLightDirection': {
			type: 'fv',
			value: []
		},
		'hemisphereLightSkyColor': {
			type: 'fv',
			value: []
		},
		'hemisphereLightGroundColor': {
			type: 'fv',
			value: []
		},
		'pointLightColor': {
			type: 'fv',
			value: []
		},
		'pointLightPosition': {
			type: 'fv',
			value: []
		},
		'pointLightDistance': {
			type: 'fv1',
			value: []
		},
		'pointLightDecay': {
			type: 'fv1',
			value: []
		},
		'spotLightColor': {
			type: 'fv',
			value: []
		},
		'spotLightPosition': {
			type: 'fv',
			value: []
		},
		'spotLightDirection': {
			type: 'fv',
			value: []
		},
		'spotLightDistance': {
			type: 'fv1',
			value: []
		},
		'spotLightAngleCos': {
			type: 'fv1',
			value: []
		},
		'spotLightExponent': {
			type: 'fv1',
			value: []
		},
		'spotLightDecay': {
			type: 'fv1',
			value: []
		}
	},
	particle: {
		'psColor': {
			type: 'c',
			value: new ColorModule.Color(15658734)
		},
		'opacity': {
			type: 'f',
			value: 1
		},
		'size': {
			type: 'f',
			value: 1
		},
		'scale': {
			type: 'f',
			value: 1
		},
		'map': {
			type: 't',
			value: null
		},
		'offsetRepeat': {
			type: 'v4',
			value: new Vector4Module.Vector4(0, 0, 1, 1)
		},
		'fogDensity': {
			type: 'f',
			value: 0.00025
		},
		'fogNear': {
			type: 'f',
			value: 1
		},
		'fogFar': {
			type: 'f',
			value: 2000
		},
		'fogColor': {
			type: 'c',
			value: new ColorModule.Color(16777215)
		}
	},
	shadowmap: {
		'shadowMap': {
			type: 'tv',
			value: []
		},
		'shadowMapSize': {
			type: 'v2v',
			value: []
		},
		'shadowBias': {
			type: 'fv1',
			value: []
		},
		'shadowDarkness': {
			type: 'fv1',
			value: []
		},
		'shadowMatrix': {
			type: 'm4v',
			value: []
		}
	}
};
exports.UniformsLib = UniformsLib;
