// three.cjs/materials/LineDashedMaterial.js
/**
 * @author alteredq / http://alteredqualia.com/
 *
 * parameters = {
 *  color: <hex>,
 *  opacity: <float>,
 *
 *  blending: THREE.NormalBlending,
 *  depthTest: <bool>,
 *  depthWrite: <bool>,
 *
 *  linewidth: <float>,
 *
 *  scale: <float>,
 *  dashSize: <float>,
 *  gapSize: <float>,
 *
 *  vertexColors: <bool>
 *
 *  fog: <bool>
 * }
 */
var ColorModule = require('../math/Color');
var MaterialModule = require('./Material');
var LineDashedMaterial;
LineDashedMaterial = function (parameters) {
	MaterialModule.Material.call(this);
	this.type = 'LineDashedMaterial';
	this.color = new ColorModule.Color(16777215);
	this.linewidth = 1;
	this.scale = 1;
	this.dashSize = 3;
	this.gapSize = 1;
	this.vertexColors = false;
	this.fog = true;
	this.setValues(parameters);
};
LineDashedMaterial.prototype = Object.create(MaterialModule.Material.prototype);
LineDashedMaterial.prototype.constructor = LineDashedMaterial;
LineDashedMaterial.prototype.clone = function () {
	var material = new LineDashedMaterial();
	MaterialModule.Material.prototype.clone.call(this, material);
	material.color.copy(this.color);
	material.linewidth = this.linewidth;
	material.scale = this.scale;
	material.dashSize = this.dashSize;
	material.gapSize = this.gapSize;
	material.vertexColors = this.vertexColors;
	material.fog = this.fog;
	return material;
};
LineDashedMaterial.prototype.isLineDashedMaterial = true;
exports.LineDashedMaterial = LineDashedMaterial;
