// three.cjs/materials/MeshFaceMaterial.js
/**
 * @author mrdoob / http://mrdoob.com/
 */
var MathModule = require('../math/Math');
var MeshFaceMaterial;
var toString = Object.prototype.toString;
MeshFaceMaterial = function (materials) {
	this.uuid = MathModule.Math.generateUUID();
	this.type = 'MeshFaceMaterial';
	this.materials = toString.call(materials).slice(8, -1) === 'Array' ? materials : [];
};
MeshFaceMaterial.prototype = {
	constructor: MeshFaceMaterial,
	toJSON: function () {
		var output = {
			metadata: {
				version: 4.2,
				type: 'material',
				generator: 'MaterialExporter'
			},
			uuid: this.uuid,
			type: this.type,
			materials: []
		};
		for (var i = 0, l = this.materials.length; i < l; i++) {
			output.materials.push(this.materials[i].toJSON());
		}
		return output;
	},
	clone: function () {
		var material = new MeshFaceMaterial();
		for (var i = 0; i < this.materials.length; i++) {
			material.materials.push(this.materials[i].clone());
		}
		return material;
	}
};
MeshFaceMaterial.prototype.isMeshFaceMaterial = true;
exports.MeshFaceMaterial = MeshFaceMaterial;
