// three.cjs/materials/MeshNormalMaterial.js
/**
 * @author mrdoob / http://mrdoob.com/
 *
 * parameters = {
 *  opacity: <float>,
 *
 *  shading: THREE.FlatShading,
 *  blending: THREE.NormalBlending,
 *  depthTest: <bool>,
 *  depthWrite: <bool>,
 *
 *  wireframe: <boolean>,
 *  wireframeLinewidth: <float>
 * }
 */
var MaterialModule = require('./Material');
var MeshNormalMaterial;
MeshNormalMaterial = function (parameters) {
	MaterialModule.Material.call(this, parameters);
	this.type = 'MeshNormalMaterial';
	this.wireframe = false;
	this.wireframeLinewidth = 1;
	this.morphTargets = false;
	this.setValues(parameters);
};
MeshNormalMaterial.prototype = Object.create(MaterialModule.Material.prototype);
MeshNormalMaterial.prototype.constructor = MeshNormalMaterial;
MeshNormalMaterial.prototype.clone = function () {
	var material = new MeshNormalMaterial();
	MaterialModule.Material.prototype.clone.call(this, material);
	material.wireframe = this.wireframe;
	material.wireframeLinewidth = this.wireframeLinewidth;
	return material;
};
MeshNormalMaterial.prototype.isMeshNormalMaterial = true;
exports.MeshNormalMaterial = MeshNormalMaterial;
