// three.cjs/materials/MeshPhongMaterial.js
/**
 * @author mrdoob / http://mrdoob.com/
 * @author alteredq / http://alteredqualia.com/
 *
 * parameters = {
 *  color: <hex>,
 *  emissive: <hex>,
 *  specular: <hex>,
 *  shininess: <float>,
 *  opacity: <float>,
 *
 *  map: new THREE.Texture( <Image> ),
 *
 *  lightMap: new THREE.Texture( <Image> ),
 *
 *  bumpMap: new THREE.Texture( <Image> ),
 *  bumpScale: <float>,
 *
 *  normalMap: new THREE.Texture( <Image> ),
 *  normalScale: <Vector2>,
 *
 *  specularMap: new THREE.Texture( <Image> ),
 *
 *  alphaMap: new THREE.Texture( <Image> ),
 *
 *  envMap: new THREE.TextureCube( [posx, negx, posy, negy, posz, negz] ),
 *  combine: THREE.Multiply,
 *  reflectivity: <float>,
 *  refractionRatio: <float>,
 *
 *  shading: THREE.SmoothShading,
 *  blending: THREE.NormalBlending,
 *  depthTest: <bool>,
 *  depthWrite: <bool>,
 *
 *  wireframe: <boolean>,
 *  wireframeLinewidth: <float>,
 *
 *  vertexColors: THREE.NoColors / THREE.VertexColors / THREE.FaceColors,
 *
 *  skinning: <bool>,
 *  morphTargets: <bool>,
 *  morphNormals: <bool>,
 *
 *	fog: <bool>
 * }
 */
var Three = require('../Three');
var Vector2Module = require('../math/Vector2');
var Vector3Module = require('../math/Vector3');
var ColorModule = require('../math/Color');
var MaterialModule = require('./Material');
var MeshPhongMaterial;
MeshPhongMaterial = function (parameters) {
	MaterialModule.Material.call(this);
	this.type = 'MeshPhongMaterial';
	this.color = new ColorModule.Color(16777215);
	// diffuse
	this.emissive = new ColorModule.Color(0);
	this.specular = new ColorModule.Color(1118481);
	this.shininess = 30;
	this.metal = false;
	this.wrapAround = false;
	this.wrapRGB = new Vector3Module.Vector3(1, 1, 1);
	this.map = null;
	this.lightMap = null;
	this.bumpMap = null;
	this.bumpScale = 1;
	this.normalMap = null;
	this.normalScale = new Vector2Module.Vector2(1, 1);
	this.specularMap = null;
	this.alphaMap = null;
	this.envMap = null;
	this.combine = Three.MultiplyOperation;
	this.reflectivity = 1;
	this.refractionRatio = 0.98;
	this.fog = true;
	this.shading = Three.SmoothShading;
	this.wireframe = false;
	this.wireframeLinewidth = 1;
	this.wireframeLinecap = 'round';
	this.wireframeLinejoin = 'round';
	this.vertexColors = Three.NoColors;
	this.skinning = false;
	this.morphTargets = false;
	this.morphNormals = false;
	this.setValues(parameters);
};
MeshPhongMaterial.prototype = Object.create(MaterialModule.Material.prototype);
MeshPhongMaterial.prototype.constructor = MeshPhongMaterial;
MeshPhongMaterial.prototype.clone = function () {
	var material = new MeshPhongMaterial();
	MaterialModule.Material.prototype.clone.call(this, material);
	material.color.copy(this.color);
	material.emissive.copy(this.emissive);
	material.specular.copy(this.specular);
	material.shininess = this.shininess;
	material.metal = this.metal;
	material.wrapAround = this.wrapAround;
	material.wrapRGB.copy(this.wrapRGB);
	material.map = this.map;
	material.lightMap = this.lightMap;
	material.bumpMap = this.bumpMap;
	material.bumpScale = this.bumpScale;
	material.normalMap = this.normalMap;
	material.normalScale.copy(this.normalScale);
	material.specularMap = this.specularMap;
	material.alphaMap = this.alphaMap;
	material.envMap = this.envMap;
	material.combine = this.combine;
	material.reflectivity = this.reflectivity;
	material.refractionRatio = this.refractionRatio;
	material.fog = this.fog;
	material.shading = this.shading;
	material.wireframe = this.wireframe;
	material.wireframeLinewidth = this.wireframeLinewidth;
	material.wireframeLinecap = this.wireframeLinecap;
	material.wireframeLinejoin = this.wireframeLinejoin;
	material.vertexColors = this.vertexColors;
	material.skinning = this.skinning;
	material.morphTargets = this.morphTargets;
	material.morphNormals = this.morphNormals;
	return material;
};
MeshPhongMaterial.prototype.isMeshPhongMaterial = true;
exports.MeshPhongMaterial = MeshPhongMaterial;
