// three.cjs/materials/SpriteMaterial.js
/**
 * @author alteredq / http://alteredqualia.com/
 *
 * parameters = {
 *  color: <hex>,
 *  opacity: <float>,
 *  map: new THREE.Texture( <Image> ),
 *
 *  blending: THREE.NormalBlending,
 *  depthTest: <bool>,
 *  depthWrite: <bool>,
 *
 *	uvOffset: new THREE.Vector2(),
 *	uvScale: new THREE.Vector2(),
 *
 *  fog: <bool>
 * }
 */
var ColorModule = require('../math/Color');
var MaterialModule = require('./Material');
var SpriteMaterial;
SpriteMaterial = function (parameters) {
	MaterialModule.Material.call(this);
	this.type = 'SpriteMaterial';
	this.color = new ColorModule.Color(16777215);
	this.map = null;
	this.rotation = 0;
	this.fog = false;
	// set parameters
	this.setValues(parameters);
};
SpriteMaterial.prototype = Object.create(MaterialModule.Material.prototype);
SpriteMaterial.prototype.constructor = SpriteMaterial;
SpriteMaterial.prototype.clone = function () {
	var material = new SpriteMaterial();
	MaterialModule.Material.prototype.clone.call(this, material);
	material.color.copy(this.color);
	material.map = this.map;
	material.rotation = this.rotation;
	material.fog = this.fog;
	return material;
};
SpriteMaterial.prototype.isSpriteMaterial = true;
exports.SpriteMaterial = SpriteMaterial;
