// three.cjs/extras/geometries/PlaneGeometry.js
/**
 * @author mrdoob / http://mrdoob.com/
 * based on http://papervision3d.googlecode.com/svn/trunk/as3/trunk/src/org/papervision3d/objects/primitives/Plane.as
 */
var PlaneBufferGeometryModule = require('./PlaneBufferGeometry');
var GeometryModule = require('../../core/Geometry');
var PlaneGeometry;
PlaneGeometry = function (width, height, widthSegments, heightSegments) {
	console.info('THREE.PlaneGeometry: Consider using THREE.PlaneBufferGeometry for lower memory footprint.');
	GeometryModule.Geometry.call(this);
	this.type = 'PlaneGeometry';
	this.parameters = {
		width: width,
		height: height,
		widthSegments: widthSegments,
		heightSegments: heightSegments
	};
	this.fromBufferGeometry(new PlaneBufferGeometryModule.PlaneBufferGeometry(width, height, widthSegments, heightSegments));
};
PlaneGeometry.prototype = Object.create(GeometryModule.Geometry.prototype);
PlaneGeometry.prototype.constructor = PlaneGeometry;
PlaneGeometry.prototype.isPlaneGeometry = true;
exports.PlaneGeometry = PlaneGeometry;
